### DISCUSSION TITLE
How to force CDATA for a specific SOAP Element?

### DISCUSSION THREAD

**Discussion by michabbb** [WsdlToPhp/PackageGenerator#297] [Q&A] [answered]
dear all,

eBay had another great idea to make developers crazy 😕 and I have no idea how to solve this with your project, so I please need some help/advise.

based on ebay WSDL I am able to generate the classmap and create a request like this:

```xml
<SOAP-ENV:Body>
        <ns1:ReviseFixedPriceItemRequest>
            <ns1:Item>
                <ns1:Description>&lt;!DOCTYPE html&gt; &lt;!--suppress HtmlFormInputWithoutLabel
                    --&gt; &lt;html&gt; &lt;head&gt; &lt;meta charset="UTF-8"/&gt;
                    &lt;title&gt;Kunde&lt;/title&gt; &lt;meta name="viewport"
                    content="width=device-width,initial-scale=1"/&gt; &lt;link rel="stylesheet"...................
                </ns1:Description>             
               </ns1:Item>
        </ns1:ReviseFixedPriceItemRequest>
    </SOAP-ENV:Body>
````

but ebay forces us (next month) to send this request like that:

```xml
<SOAP-ENV:Body>
    <ns1:ReviseFixedPriceItemRequest>
        <ns1:Item>
            <ns1:Description>
                <![CDATA[
                    <!DOCTYPE html> <!--suppress HtmlFormInputWithoutLabel --> <html> <head> <meta charset="UTF-8"/>
                    <title>Kunde</title> <meta name="viewport" content="width=device-width,initial-scale=1"/> <link rel="stylesheet"...................
                ]]>
            </ns1:Description>            
        </ns1:Item>
    </ns1:ReviseFixedPriceItemRequest>
</SOAP-ENV:Body>
```

and sadly I have absolutely no clue how to make this work.
i would be very happy, if you could help me here !! 🙏 🙏 

[here's the latest trading WSDL of ebay](https://developer.ebay.com/webservices/latest/ebaysvc.wsdl).

thanks !!! 
micha

**Comment by mikaelcom** [comment]
Hi,
I had to comment SeeLink and CallName tags but I still have issues to load the WSDL with the SoapClient class. If you have a working WSDL, please send it back

**Reply by michabbb** [reply]
thanks! i remeber,  i had to remove that stuff, too. let me check....

**Reply by michabbb** [reply]
@mikaelcom please try [this](https://github.com/michabbb/php-ebay-sdk/blob/php8/wsdl/trading.wsdl) one.

**Comment by michabbb** [comment, marked as answer]
the WSDL generator creates a file called "ItemType.php" which contains this method:

```php
public function setDescription(?string $description = null): self
    {
        // validation for constraint: string
        if (!is_null($description) && !is_string($description)) {
            throw new InvalidArgumentException(sprintf('Invalid value %s, please provide a string, %s given', var_export($description, true), gettype($description)), __LINE__);
        }
       $this->Description = $description;
        return $this;
    }
```

I tried now this hack:

```php
public function setDescription(?string $description = null): self
    {
        // validation for constraint: string
        if (!is_null($description) && !is_string($description)) {
            throw new InvalidArgumentException(sprintf('Invalid value %s, please provide a string, %s given', var_export($description, true), gettype($description)), __LINE__);
        }
        $this->Description = new \SoapVar('<ns1:Description><![CDATA['.$description.']]></ns1:Description>', XSD_ANYXML);
        return $this;
    }
```

this generates this request:

```xml
<SOAP-ENV:Body>
  <ns1:ReviseItemRequest>
    <ns1:Item>
      <ns1:Description>
          <![CDATA[test description]]>
       </ns1:Description>
    </ns1:Item>
  </ns1:ReviseItemRequest>
</SOAP-ENV:Body>
```

I don´t really like this "hack" because:

- I have to put in the XML tag `<ns1:Description>` 
- I don´t see any way to tell the generator to create exactly this code
- I am forced the change types of `$Description` in "ItemType.php":

```php
// from this
/*
 * @var string|null
 */
    protected ?string $Description = null;

// to this
/*
 * @var string|null|SoapVar
 */
    protected $Description = null;
```


@mikaelcom in case you have a better idea or solution for this, please let me know 🙏 😞

**Reply by mikaelcom** [reply]
Sincerely, if you only have one property of only one class, I think it's the best way. Let me know otherwise.

**Reply by michabbb** [reply]
@mikaelcom is there any other way? I mean, do you know any alternative to this? And is there anything in this generator that lets me customize the generated class so that I don't have to manually change it AFTER is has been generated???

**Reply by mikaelcom** [reply]
The other way is to override the SoapClient class used by the generated classes in order to rewrite the XML request before it is actually sent. Doing it for just one property can seem overkill but it should work :
- the option to use your own SoapClient implementation is https://github.com/WsdlToPhp/PackageGenerator/wiki/Options#soapclient
- sample code can be found within https://github.com/WsdlToPhp/PackageEws365/tree/develop/SoapClient
Good luck

**Reply by michabbb** [reply]
okay, thanks for letting me know!
